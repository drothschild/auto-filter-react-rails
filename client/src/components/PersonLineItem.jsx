import React from 'react';

const PersonLineItem = ({ person }) => (
  <div className="person-line-item">
    {person.attributes.firstName} {person.attributes.lastName}
  </div>
);

export default PersonLineItem;
