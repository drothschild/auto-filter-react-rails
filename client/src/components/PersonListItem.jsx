import React from 'react';

const PersonListItem = Person => (
  <div className="person-line-item">
    {Person.attributes.firstName} {Person.attributes.lastName}
  </div>
);

export default PersonListItem;
