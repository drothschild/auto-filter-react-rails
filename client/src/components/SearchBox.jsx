import React, { Component } from 'react';
import { connect } from 'react-redux';
import { changeSearchTerm } from '../redux/actionCreators';

class SearchBox extends Component {
  constructor(props) {
    super(props);
    this.handleSearchTermChange = this.handleSearchTermChange.bind(this);
  }
  handleSearchTermChange(event) {
    const term = event.target.value;
    const loc = (term === '') ? '' : `?name=${encodeURIComponent(term)}`;
    window.history.pushState('', '', `${window.location.origin}/search${loc}`);
    this.props.dispatchSearchTermChange(term);
  }
  render() {
    const { searchTerm } = this.props;
    return (
      <div className="search-box">
        <input
          type="text"
          placeholder="Search"
          value={searchTerm}
          onChange={this.handleSearchTermChange}
        />
      </div>
    );
  }
}

SearchBox.defaultProps = {
  searchTerm: '',
  dispatchSearchTermChange: () => '',
};

const mapStateToProps = state => ({ searchTerm: state.searchTerm });

const mapDispatchToProps = dispatch => ({
  dispatchSearchTermChange(term) {
    dispatch(changeSearchTerm(term));
  },
});

export const Unwrapped = SearchBox;
export default connect(mapStateToProps, mapDispatchToProps)(SearchBox);
