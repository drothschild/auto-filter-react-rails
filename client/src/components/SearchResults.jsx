import React from 'react';
import { connect } from 'react-redux';
import PersonLineItem from './PersonLineItem';

const SearchResults = ({ searchTerm, people, isFetching }) => {
  const possibleWaitingAnnouncement = isFetching ? (
    <div className="is-fetching">Fetching Data...</div>
  ) : null;
  const peopleList = people
    .filter(person =>
      `${person.attributes.firstName} ${person.attributes.lastName}`
        .toUpperCase()
        .indexOf(searchTerm.toUpperCase()) >= 0)
    .sort((a, b) => {
      if (a.attributes.lastName < b.attributes.lastName) {
        return -1;
      }
      return 1;
    })
    .map(person => <PersonLineItem person={person} key={person.id} />);
  return (
    <div className="search-results">
      {possibleWaitingAnnouncement}
      {peopleList}
    </div>
  );
};

SearchResults.defaultProps = {
  people: [],
  searchTerm: '',
  isFetching: true,
};

const mapStateToProps = state => ({
  people: state.people.data,
  searchTerm: state.searchTerm,
  isFetching: state.people.isFetching,
});


export const Unwrapped = SearchResults;

export default connect(mapStateToProps)(SearchResults);
