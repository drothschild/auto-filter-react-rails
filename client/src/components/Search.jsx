import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router';
import queryString from 'query-string';
import SearchBox from './SearchBox';
import SearchResults from './SearchResults';
import { fetchPeople, changeSearchTerm } from '../redux/actionCreators';

class Search extends Component {
  componentDidMount() {
    const {
      location, dispatchSearchTermChange, people, getPeople,
    } = this.props;
    if (location.search) {
      const parsed = queryString.parse(location.search);
      if (parsed.name) {
        dispatchSearchTermChange(parsed.name);
      }
    }
    if (people.length === 0) {
      getPeople();
    }
  }
  render() {
    return (
      <div>
        <SearchBox />
        <SearchResults />
      </div>
    );
  }
}

Search.defaultProps = {
  people: [],
  getPeople: () => '',
  dispatchSearchTermChange: () => '',
};

const mapStateToProps = state => ({
  people: state.people.data,
});

const mapDispatchToProps = dispatch => ({
  getPeople() {
    dispatch(fetchPeople());
  },
  dispatchSearchTermChange(term) {
    dispatch(changeSearchTerm(term));
  },
});
export const Unwrapped = Search;

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(Search));
