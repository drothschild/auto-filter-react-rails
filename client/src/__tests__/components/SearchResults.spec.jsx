import React from 'react';
import { configure, shallow, render } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import { Unwrapped as UnwrappedSearchResults } from '../../components/SearchResults';
import People from '../dummydata.json';

configure({ adapter: new Adapter() });

test('It should render search results', () => {
  const component = render(<UnwrappedSearchResults people={People} />);
  expect(component.find('.person-line-item')).toHaveLength(People.length);
});

test('It should filter first name with the search term', () => {
  const component = render(<UnwrappedSearchResults people={People} searchTerm="Jane" />);
  expect(component.find('.person-line-item')).toHaveLength(1);
});

test('It should filter last name with the search term', () => {
  const component = render(<UnwrappedSearchResults people={People} searchTerm="Doe" />);
  expect(component.find('.person-line-item')).toHaveLength(1);
});
test('It should filter out all records if no matching search term', () => {
  const component = render(<UnwrappedSearchResults people={People} searchTerm="zzz" />);
  expect(component.find('.person-line-item')).toHaveLength(0);
});

test('It should show waiting status', () => {
  const component = shallow(<UnwrappedSearchResults isFetching />);
  expect(component.find('.is-fetching')).toHaveLength(1);
});

test('It should hide waiting status', () => {
  const component = shallow(<UnwrappedSearchResults isFetching="false" />);
  expect(component.find('.is-fetching')).toHaveLength(1);
});

test('By default, it should show no people and a waiting status', () => {
  const component = shallow(<UnwrappedSearchResults />);
  expect(component.find('.person-line-item')).toHaveLength(0);
  expect(component.find('.is-fetching')).toHaveLength(1);
});
