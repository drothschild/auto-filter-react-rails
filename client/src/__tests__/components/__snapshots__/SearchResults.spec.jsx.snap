// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SearchResults
    isFetching={true}
    people={Array []}
    searchTerm=""
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="is-fetching"
>
          Fetching Data...
</div>,
        Array [],
      ],
      "className": "search-results",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Fetching Data...",
          "className": "is-fetching",
        },
        "ref": null,
        "rendered": "Fetching Data...",
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="is-fetching"
>
            Fetching Data...
</div>,
          Array [],
        ],
        "className": "search-results",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Fetching Data...",
            "className": "is-fetching",
          },
          "ref": null,
          "rendered": "Fetching Data...",
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
